src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1508 @@
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTRUSIVELIST_HPP
+#define SHARE_UTILITIES_INTRUSIVELIST_HPP
+
+#include "memory/allStatic.hpp"
+#include "metaprogramming/enableIf.hpp"
+#include "metaprogramming/logical.hpp"
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+#include <type_traits>
+
+class IntrusiveListEntry;
+class IntrusiveListImpl;
+
+/**
+ * The IntrusiveList class template provides a doubly-linked list in
+ * which the links between elements are embedded directly into objects
+ * contained in the list.  As a result, there are no copies involved
+ * when inserting objects into the list or referencing list objects,
+ * and removing an object from a list need not involve destroying the
+ * object.
+ *
+ * To be used in a IntrusiveList, an object must have a
+ * IntrusiveListEntry member.  An IntrusiveList is associated with the
+ * class of its elements and the entry member.
+ *
+ * An object can be in multiple lists at the same time, so long as
+ * each list uses a different entry member.  That is, the class of the
+ * object must have multiple IntrusiveListEntry members, one for each
+ * list the object is simultaneously an element.
+ *
+ * The API for IntrusiveList is modelled on the containers provided by
+ * the C++ standard library.  In particular, iteration over the
+ * elements is provided via iterator classes.
+ *
+ * IntrusiveLists support polymorphic elements.  Because the objects
+ * in a list are externally managed, rather than being embedded values
+ * in the list, the actual type of such objects may be more specific
+ * than the list's element type.
+ *
+ * * T is the class of the elements in the list.  Must be a possibly
+ * const-qualified class type.
+ *
+ * * entry_member is a pointer to class member referring to the
+ * IntrusiveListEntry subobject of T used by this list.
+ *
+ * * has_size determines whether the list has a size()
+ * operation, returning the number of elements in the list.  If the
+ * operation is present, it has constant-time complexity.  The default
+ * is to not provide a constant-time size() operation.
+ *
+ * * Base is the base class for the list.  This is typically
+ * used to specify the allocation class.  The default is void, indicating
+ * no allocation class for the list.
+ *
+ * The value type for a list may be const-qualified.  Such a list provides
+ * const iterators and access to const-qualified elements.  A list whose value
+ * type is not const-qualified can provide both const and non-const iterators
+ * and can provide access to either const or unqualified elements, depending
+ * on the operation used.  A list that is itself const only provides const
+ * iterators and access to const-qualified elements.  A const object cannot be
+ * added to a list whose value type is not const-qualified, as that would be
+ * an implicit casting away of the const qualifier.
```

Comment:
Okay, I feel like this can be shortened signifcantly:

> A const-qualified type can be part of an IntrusiveList, then you only get const iterators and access to const elements. If you use a non-const type, then you can get both const and non-const iterators and access to elements. You can't add const values to a non-const list, as that would be implicitly casting away the const qualifier.

Reply:
That rewrite isn't correct.

You _can_ add const values to a non-const list. In fact, you can only add
values (const or not) to a non-const list. A const list has type `const
IntrusiveList<T, accessor>`, while a non-const list is similar but without the
`const` qualifier.

What you can't do is add a const value to a (necessarily non-const) list of
non-const elements.  So if T is an unqualified class type, then you can add
const values to an `IntrusiveList<const T, accessor>` but not to an
`IntrusiveList<T, accessor>`.

Reply:
Is something like this more clear? In my experience, non-intrusive collections
with const-qualified elements are uncommon, so I found the implications here
not immediately obvious.

```
 * A const iterator has a const-qualified element type, and provides const
 * access to the elements of the associated list.  A non-const iterator has an
 * unqualified element type, and provides mutable element access.  A non-const
 * iterator is implicitly convertible to a corresponding const iterator.
 *
 * A const list provides const iterators and access to const-qualified
 * elements, and cannot be used to modify the sequence of elements.  Only a
 * non-const list can be used to modify the sequence of elements.
 *
 * A list can have a const-qualified element type, providing const iterators
 * and access to const-qualified elements.  A const object cannot be added to
 * a list with an unqualified element type, as that wuold be an implicit
 * casting away of the const qualifier.
```

Reply:
Nit typo: s/wuold/would/

Reply:
My suggestion isn't quite right either; the nomenclature for a list/iterator with const elements needs a new term.
I've pushed another rewrite, defining the terms "const-element list/iterator" and using that terminology elsewhere.